File: v1/src/main/java/com/api/v1/persons/services/PersonModificationServiceImpl.java
package com.api.v1.persons.services;

import com.api.v1.persons.domain.Person;
import com.api.v1.persons.domain.PersonAuditTrail;
import com.api.v1.persons.domain.PersonAuditTrailRepository;
import com.api.v1.persons.domain.PersonRepository;
import com.api.v1.persons.dtos.PersonModificationRequestDto;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class PersonModificationServiceImpl implements PersonModificationService {

    @Autowired
    private PersonRepository personRepository;

    @Autowired
    private PersonAuditTrailRepository auditTrailRepository;

    @Override
    public Mono<Person> modify(@NotNull Person person, @Valid PersonModificationRequestDto requestDto) {
        return auditTrailRepository
                .save(PersonAuditTrail.of(person))
                .then(Mono.defer(() -> {
                    person.modify(requestDto);
                    return personRepository.save(person);
                }));
    }

}

File: v2/src/main/kotlin/com/api/v2/persons/services/PersonModificationServiceImpl.kt
package com.api.v2.persons.services

import com.api.v2.persons.domain.Person
import com.api.v2.persons.domain.PersonAuditTrail
import com.api.v2.persons.domain.PersonAuditTrailRepository
import com.api.v2.persons.domain.PersonRepository
import com.api.v2.persons.dtos.PersonModificationDto
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class PersonModificationServiceImpl: PersonModificationService {

    @Autowired
    lateinit var personRepository: PersonRepository

    @Autowired
    lateinit var personAuditTrailRepository: PersonAuditTrailRepository

    override suspend fun modify(
        person: @NotNull Person,
        requestModificationDto: @Valid PersonModificationDto
    ): Person {
        return withContext(Dispatchers.IO) {
            val personAuditTrail = PersonAuditTrail.of(person)
            val savedAuditTrail = personAuditTrailRepository.save(personAuditTrail)
            person.modify(requestModificationDto)
            personRepository.save(person)
        }
    }

}
